goss: init at 0.3.16

Hyzual opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Closes #121496

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

legendofmiracles commented on maintainers/maintainer-list.nix:
```diff
@@ -4121,6 +4121,12 @@
     githubId = 12491746;
     name = "Masato Yonekawa";
   };
+  hyzual = {
+    email = "hyzual@gmail.com";
+    github = "Hyzual";
+    githubId = 2051507;
+    name = "Joris Masson";
+  };
```
Adding maintainers is usually in another commit `maintainers: add Joris Masson`

Hyzual replied:
Oops sorry about that 😅. I'll split into two commits. Thanks for the review 😊

legendofmiracles commented:
Result of `nixpkgs-review pr 121497` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>goss</li>
  </ul>
</details>

Hyzual committed:
maintainers: add Joris Masson

Hyzual committed:
goss: init at 0.3.16

Co-authored-by: j-k <dev@j-k.io>

06kellyjac commented on pkgs/tools/misc/goss/default.nix:
```diff
@@ -0,0 +1,30 @@
+
+  vendorSha256 = "1lyqjkwj8hybj5swyrv6357hs8sxmf4wim0c8yhfb9mv7fsxhrv7";
+
+  buildFlagsArray = [ "-ldflags=-s -w -X=main.version=${version}" ];
```
```suggestion
  preBuild = ''
    buildFlagsArray+=("-ldflags" "-s -w -X main.version=v${version}")
  '';
```

It's arguably better to set `buildFlagsArray` through bash currently, but it's not a big deal, especially for just setting ldflags.
Go packaging docs also don't reflect this suggestion atm.
More details here:
https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/no-flags-array.md

Hyzual replied:
Thanks for pointing that out, I'll give `nixpkgs-hammering` a closer look :)

06kellyjac commented on pkgs/tools/misc/goss/default.nix:
```diff
@@ -0,0 +1,30 @@
+
+  src = fetchFromGitHub {
+    owner = "aelsabbahy";
+    repo = "goss";
```
```suggestion
    repo = pname;
```

Can use `pname` here or just leave it as-is

Hyzual replied:
Nice !

06kellyjac reviewed (approved):
Looks good, welcome to nixpkgs @Hyzual! :tada: 

I added goss to my nur a couple days ago but I've been lazy in not adding it upstream :sweat_smile: 
https://github.com/06kellyjac/nur-packages/blob/master/pkgs/goss/default.nix

I've made a couple suggestions that you're welcome to implement or disregard :slightly_smiling_face:

Hyzual commented on pkgs/tools/misc/goss/default.nix:
```diff
@@ -0,0 +1,30 @@
+
+  buildFlagsArray = [ "-ldflags=-s -w -X=main.version=${version}" ];
+
+  meta = {
+    homepage = "https://github.com/aelsabbahy/goss/";
+    description = "Quick and easy server validation";
+    longDescription = ''
+      Goss is a YAML based serverspec alternative tool for validating a server’s configuration.
+      It eases the process of writing tests by allowing the user to generate tests from the current system state.
+      Once the test suite is written they can be executed, waited-on, or served as a health endpoint.
+    '';
+    license = lib.licenses.asl20;
+    platforms = with lib.platforms; linux;
+    maintainers = with lib.maintainers; [ hyzual ];
+  };
```
I'll add `with lib` to the `meta`. I think I'll leave platforms at linux for now, I have no way of testing if it works for darwin.
Thanks for the offer, I'll add you to the maintainer list :)

06kellyjac commented:
Result of `nixpkgs-review pr 121497` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>goss</li>
  </ul>
</details>

06kellyjac reviewed (approved):
LGTM

Thanks for squashing the changes :+1:

Hyzual commented:
The `CGO_ENABLED=0` env var was not present, which resulted in a dynamically-linked binary. I added it, so now the build should match the [release script](https://github.com/aelsabbahy/goss/blob/master/release-build.sh) used in goss repository.
I also added the `changelog` meta attribute you had suggested, I had forgotten it :sweat_smile:.

Result of `nixpkgs-review pr 121497` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>goss</li>
  </ul>
</details>

06kellyjac commented:
I'm going to be a bit overly verbose with my thoughts here to give you an idea what I was thinking:
Why does it need to be static:
I can understand the goss release has to be static so it can be copied into containers just fine and used on a range of machines but due to the way nixpkgs builds things a dynamically linked binary is essentially static:

```
λ ldd ./result/bin/goss
	linux-vdso.so.1 (0x00007fff8f2da000)
	libpthread.so.0 => /nix/store/v8q6nxyppy1myi3rxni2080bv8s9jxiy-glibc-2.32-40/lib/libpthread.so.0 (0x00007f0c92560000)
	libc.so.6 => /nix/store/v8q6nxyppy1myi3rxni2080bv8s9jxiy-glibc-2.32-40/lib/libc.so.6 (0x00007f0c9239f000)
	/nix/store/v8q6nxyppy1myi3rxni2080bv8s9jxiy-glibc-2.32-40/lib/ld-linux-x86-64.so.2 => /nix/store/v8q6nxyppy1myi3rxni2080bv8s9jxiy-glibc-2.32-40/lib64/ld-linux-x86-64.so.2 (0x00007f0c92583000)
```

I guess we can expect nix users might also want to copy their local goss into containers rather than copying in a static one off github (in which case the `/nix/store` links wont exist)
And I guess the `dgoss` and `kgoss` scripts work off this principal 
https://github.com/aelsabbahy/goss/blob/master/extras/dgoss/dgoss#L77-L78
https://github.com/aelsabbahy/goss/blob/master/extras/kgoss/kgoss#L53-L64

TL;DR in this case it's a good reason to also statically compile, but it's not always necessary to copy the upstream release exactly

Also in this case (and with many go binaries) it's actually slightly smaller when statically compiled because it doesn't really link to much: `10580` before -> `10540` after

Hyzual commented:
Indeed, I made this change after trying to package `dgoss` which depends on `goss` being statically linked (otherwise we get a `not found` error).

06kellyjac commented:
/marvin opt-in
/status needs_reviewer

Hyzual commented:
Thanks for the merge :)
